marlowe-object/src/Language/Marlowe/Object/Bundler.hs

```diff
@@ -0,0 +1,231 @@
+
+instance (MonadBundler m, Monoid w) => MonadBundler (WriterT w m) where
+  define = lift . define
+
+instance (MonadBundler m) => MonadBundler (MaybeT m) where
+  define = lift . define
+
+instance (MonadBundler m) => MonadBundler (ExceptT e m) where
+  define = lift . define
+
+instance (MonadBundler m) => MonadBundler (ContT r m) where
+  define = lift . define
+
+instance (MonadBundler m) => MonadBundler (Proxy a' a b' b m) where
+  define = lift . define
+
+instance (Functor m) => MonadBundler (BundlerT m) where
+  define obj = Define obj $ Pure ()
+
+instance (Functor m) => Applicative (BundlerT m) where
+  pure = Pure
+  bf <*> ba = go bf
+    where
+      go = \case
+        Pure f -> f <$> ba
+        M m -> M (go <$> m)
+        Define obj bf' -> Define obj $ go bf'
+
+instance (Functor m) => Monad (BundlerT m) where
+  b >>= f = go b
+    where
+      go = \case
+        Pure a -> f a
+        M m -> M (go <$> m)
+        Define obj b' -> Define obj $ go b'
+
+instance (Functor m) => MonadZip (BundlerT m) where
+  mzip = liftA2 (,)
+
+instance (Functor m, Semigroup a) => Semigroup (BundlerT m a) where
+  (<>) = liftA2 (<>)
+
+instance (Functor m, Monoid a) => Monoid (BundlerT m a) where
+  mempty = pure mempty
+
+instance MonadTrans BundlerT where
+  lift = M . fmap pure
+
+instance (MonadFail m) => MonadFail (BundlerT m) where
+  fail = lift . fail
+
+instance (MonadIO m) => MonadIO (BundlerT m) where
+  liftIO = lift . liftIO
+
+instance (MonadReader r m) => MonadReader r (BundlerT m) where
+  ask = lift ask
+  local f = go
+    where
+      go = \case
+        Pure a -> Pure a
+        M m -> M $ go <$> local f m
+        Define obj b -> Define obj $ go b
+
+instance (MonadState s m) => MonadState s (BundlerT m) where
+  get = lift get
+  put = lift . put
+  state = lift . state
+
+instance (MonadWriter w m) => MonadWriter w (BundlerT m) where
+  writer = lift . writer
+  tell = lift . tell
+  listen = go mempty
+    where
+      go w = \case
+        Pure a -> Pure (a, w)
+        M m -> M do
+          (b, w') <- listen m
+          let !w'' = w <> w'
+          pure $ go w'' b
+        Define obj b -> Define obj $ go w b
+  pass = go mempty
+    where
+      go w = \case
+        Pure (a, f) -> M $ pass $ pure (Pure a, const $ f w)
+        M m -> M do
+          (b, w') <- censor (const mempty) $ listen m
+          let !w'' = w <> w'
+          pure $ go w'' b
+        Define obj b -> Define obj $ go w b
+
+instance (MonadError e m) => MonadError e (BundlerT m) where
+  throwError = lift . throwError
+  catchError b f = go b
+    where
+      go = \case
+        Pure a -> Pure a
+        M m -> M $ (go <$> m) `catchError` (pure . f)
+        Define obj b' -> Define obj $ go b'
+
+-- | Run a bundler action, returning the resulting object bundle and the result of the action.
+runBundlerT :: (Monad m) => BundlerT m a -> m (ObjectBundle, a)
+runBundlerT = (fmap . first) ObjectBundle . toListM' . runBundlerTIncremental
```

Comment:
To use of `toListM'` here hints to reconsider having the `[]` in `ObjectBundle`. Could be replace by a `Producer`, couldn't it?

Reply:
I suppose it could, but that would be a pretty major redesign of `ObjectBundle` and its interface at this point